File: 2018/06-nearest-neighbour-coordinates/go/star1.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strings"
)

type point struct {
	x int
	y int
}

func absInt(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func distance(p1, p2 point) int {
	return absInt(p1.x-p2.x) + absInt(p1.y-p2.y)
}

func main() {
	bytes, _ := ioutil.ReadFile("input")
	lines := strings.Split(string(bytes), "\n")

	topLeft := point{math.MaxInt32, math.MaxInt32}
	bottomRight := point{math.MinInt32, math.MinInt32}

	var coords []point
	for _, line := range lines {
		if line == "" {
			continue
		}

		var x, y int
		fmt.Sscanf(line, "%d, %d", &x, &y)
		coords = append(coords, point{x, y})

		topLeft = point{minInt(topLeft.x, x), minInt(topLeft.y, y)}
		bottomRight = point{maxInt(bottomRight.x, x), maxInt(bottomRight.y, y)}
	}

	areas := make(map[point]int)
	maxArea := 0
	for x := topLeft.x + 1; x < bottomRight.x-1; x++ {
		for y := topLeft.y + 1; y < bottomRight.y-1; y++ {
			closestToOnePoint := true
			closest := point{math.MaxInt32, math.MaxInt32}
			for _, coord := range coords {
				dist := distance(coord, point{x, y})

				if dist < distance(closest, point{x, y}) {
					closest = coord
					closestToOnePoint = true
				} else if dist == distance(closest, point{x, y}) {
					closestToOnePoint = false
				}
			}

			if closestToOnePoint {
				areas[closest] += 1
				maxArea = maxInt(maxArea, areas[closest])
			}
		}
	}

	fmt.Println(maxArea)
}

File: 2018/06-nearest-neighbour-coordinates/rust/star1.rs
use std::collections::HashMap;

const INPUT: &str = include_str!("../input");

fn distance(point1: (i32, i32), point2: (i32, i32)) -> i32 {
    (point1.0 - point2.0).abs() + (point1.1 - point2.1).abs()
}

fn parse_line(line: &str) -> (i32, i32) {
    let ints: Vec<i32> = line.split(", ").map(|s| s.parse().unwrap()).collect();

    (ints[0], ints[1])
}

fn main() {
    let coordinates: Vec<(i32, i32)> = INPUT.lines().map(|l| parse_line(l)).collect();

    let top_left = (
        coordinates.iter().map(|c| c.0).min().unwrap(),
        coordinates.iter().map(|c| c.1).min().unwrap(),
    );
    let bottom_right = (
        coordinates.iter().map(|c| c.0).max().unwrap(),
        coordinates.iter().map(|c| c.1).max().unwrap(),
    );

    let mut areas: HashMap<(i32, i32), i32> = HashMap::new();
    for x in (top_left.0 + 1)..(bottom_right.0 - 1) {
        for y in (top_left.1 + 1)..(bottom_right.1 - 1) {
            let mut distances: Vec<(i32, (i32, i32))> = coordinates
                .iter()
                .map(|c| (distance(*c, (x, y)), *c))
                .collect();

            distances.sort();
            if distances[0].0 != distances[1].0 {
                let closest = distances[0].1;
                *areas.entry(closest).or_default() += 1;
            }
        }
    }

    let largest_area = areas.values().max().unwrap();
    println!("{}", largest_area);
}
